concourse/client/util.py

```diff
@@ -122,3 +137,118 @@ def jobs_not_triggered(
+
+
+def process_untriggered_job(
+    job: Job,
+    resource: PipelineConfigResource,
+    resource_version: ResourceVersion,
+    concourse_api,
+    retries: int=3,
+):
+    while retries >= 0:
+        try:
+            with pin_resource(
+                resource=resource,
+                resource_version=resource_version,
+                concourse_api=concourse_api,
+            ):
+                concourse_api.trigger_build(
+                    pipeline_name=resource.pipeline_name(),
+                    job_name=job.pipeline_name,
+                )
+
+                if wait_for_job_to_be_triggered(
+                    job=job,
+                    resource_version=resource_version,
+                    concourse_api=concourse_api,
+                    retries=10,
+                    sleep_time_seconds=5
+                ):
+                    info(
+                        f'job {job.name} for resource version {resource_version.version()} '
+                        'has been triggered'
+                    )
+                else:
+                    warning(
+                        f'job {job.name} for resource version {resource_version.version()} '
+                        'could not be triggered'
+                    )
+                return
+        except PinningFailed:
+            retries -= 1
+
+    warning(f'Could not process untriggered job {job.name}')
+
+
+@contextmanager
+def pin_resource(
+    resource: PipelineResource,
+    resource_version: ResourceVersion,
+    concourse_api,
+):
+    pin_comment = {
+        'datetime': datetime.now().isoformat(),
+        'resource_type': resource.type().value,
+        'version_data': resource_version.version(),
+        'comment': 'automatically pinned by technical user, please do not unpin manually',
+    }
+    _ensure_resource_unpinned(
+        resource=resource,
+        concourse_api=concourse_api,
+        retries=10,
+        sleep_time_seconds=5,
+    )
+    try:
+        concourse_api.pin_and_comment_resource_version(
+            pipeline_name=resource.pipeline_name(),
+            resource_name=resource.name,
+            resource_version_id=resource_version,
+            comment=json.dumps(pin_comment),
+        )
+        yield
+    finally:
+        concourse_api.unpin_resource(
+            pipeline_name=resource.pipeline_name(),
+            resource_name=resource.name,
+        )
+
+
+def _ensure_resource_unpinned(
+    resource: PipelineResource,
+    concourse_api,
+    retries: int=10,
+    sleep_time_seconds: int=5,
+):
+    while retries >= 0:
+        cc_resource = concourse_api.resource(
+            pipeline_name=resource.pipeline_name(),
+            resource_name=resource.name,
+        )
+
+        if not cc_resource.is_pinned():
+            return
+
+        pin_comment = cc_resource.pin_comment()
+        try:
+            # if comment is json format, we assume it was pinned by us
```

Comment:
please do not only take into account the fact whether or not it is json, but rather also check for the attributes. it could be that case that someone else write some different json